[DX] ADR usage

Guikingone opened the pull request:
With the new approach used by the container, the old controller can be easily replaced by the ADR pattern, as Symfony grow, I think it should be the role of the documentation to show how to use the pattern and how to adapt the framework structure.

PS : I'm not pretty sure of the 'note' syntax.

Guikingone committed:
Adding the ADR approach. 

With the new approach used by the container, the old controller can be replaced easily by the ADR pattern, as Symfony grow, I think it should be the role of the documentation to show how to use the pattern and how to adapt the framework structure.

PS : I'm not pretty sure of the 'note' syntax.

Guikingone committed:
Update service.rst

Remove the PHP 7.1 syntax.

Guikingone committed:
Update service.rst

Removed the ..note block.

yceruto reviewed (changes requested):
This one documents part of https://github.com/symfony/symfony/pull/21723 so we may add a note for this feature `@Route("/hello", name="hello")` explaining when this only work? and close https://github.com/symfony/symfony-docs/issues/7549 then ;) Thanks!

yceruto commented on controller/service.rst:
```diff
@@ -80,6 +80,118 @@ If your controller implements the ``__invoke()`` method - popular with the
+        # app/config/routing.yml
+        hello:
+            path:     /hello
+            methods: ['GET']
```
This requirement has not been configured in the rest of formats so I guess can be removed?

yceruto commented on controller/service.rst:
```diff
@@ -80,6 +80,118 @@ If your controller implements the ``__invoke()`` method - popular with the
+        {
+            return new Response(
+                $this->renderEngine->render('default/index.html.twig')
+            );
```
should be in one line

yceruto commented on controller/service.rst:
```diff
@@ -80,6 +80,120 @@ If your controller implements the ``__invoke()`` method - popular with the
+         * @param EngineInterface $renderEngine
+         */
+        public function __construct(
+            EngineInterface $renderEngine
```
Maybe just `engine`?  this do a bit more than only `render` ;)

yceruto reviewed (changes requested):
A few more... ;)

Guikingone committed:
Update service.rst

Guikingone commented:
@yceruto For the docblocks, yes, for sure, I've updated the commits. 

For the code, I'm not in the same position, if you look at the PSR (especially the 2 http://www.php-fig.org/psr/psr-2/#overview), they say clearly that a line should be under 80: 

- There MUST NOT be a hard limit on line length; the soft limit MUST be 120 characters; lines SHOULD be 80 characters or less.

If you pass the __construct in one line, yes, that works but for the Response return, that's over 85 characters long, way too much in my opinion. 

For the config files, that's a personal opinion but the requirements key MUST be mandatory and in fact, Insight asks for this key if she's not defined, in my opinion, the documentation should show the right way of using the files and classes.

For the parameters key, I think she can stay for simple visibility and easier understanding of the whole file but in the "simplest" way of configuring the service, yes, she can be removed.

For the PR, I'm not sure of understanding your vision :/

yceruto commented on controller/service.rst:
```diff
@@ -80,6 +80,103 @@ If your controller implements the ``__invoke()`` method - popular with the
+        public function __construct(
+            EngineInterface $engine
+        ) {
+            $this->renderEngine = $render;
```
`$this->engine = $engine` ? and fix related use

yceruto commented on controller/service.rst:
```diff
@@ -80,6 +80,103 @@ If your controller implements the ``__invoke()`` method - popular with the
+        
+        public function __construct(
+            EngineInterface $engine
+        ) {
```
`public function __construct(EngineInterface $engine)` one line?

Guikingone committed:
Update service.rst

Guikingone committed:
Update service.rst

Guikingone commented:
@yceruto Should be better.

yceruto commented on controller/service.rst:
```diff
@@ -80,6 +80,118 @@ If your controller implements the ``__invoke()`` method - popular with the
+            methods: ['GET']
+            defaults: { _controller: AppBundle\Action\HelloAction }
+
+    .. code-block:: php-annotations
```
http://symfony.com/doc/current/contributing/documentation/standards.html#formats:
> * Routing: Annotations, YAML, XML, PHP

yceruto commented:
Please, I'm only address you to [Coding Standards for Symfony project][1] (i.e these aren't my personal preferences).

> For the code, I'm not in the same position, if you look at the PSR (especially the 2 http://www.php-fig.org/psr/psr-2/#overview), they say clearly that a line should be under 80

Following http://symfony.com/doc/current/contributing/code/standards.html#structure:
> Declare all the arguments on the same line as the method/function name, no matter how many arguments there are;

That's applicable to:
* https://github.com/symfony/symfony-docs/pull/8153#discussion_r127471291
* https://github.com/symfony/symfony-docs/pull/8153#discussion_r127485496

  [1]: http://symfony.com/doc/current/contributing/code/standards.html

Guikingone committed:
Update service.rst

Guikingone commented:
Should be better, I understand your vision, I continue to use the PSR rules into my projects, that's why I recommend the multi-lines approach :). 

For the annotations panels order, I've just followed the order used in the first part of the document, no more :)

Guikingone committed:
Update service.rst

Change the order of the routing files and deleted the methods key in yaml routing.

Guikingone commented:
@yceruto Should be better, I've removed the methods key and change the order of the recommended routing configuration. 

For the parameters key, I think about using '//' in order add a reminder that some keys are missing from the example but that's just for the example, what do you think of this approach? 

```yml

// ...

    services:
         _defaults:
             autowire: true
             autoconfigure: true
             public: false
             
         # Allow to load every actions
         AppBundle\Action\:
             resource: '../../src/AppBundle/Action/'
             public: true
```

Guikingone committed:
Update service.rst

Guikingone commented:
@yceruto Services.yml corrected, I've use the # ... syntax, Yaml doesn't support // ..., I've thought about this after my comments :(

Guikingone committed:
Update service.rst

Guikingone commented:
@yceruto Hi, small question about this PR, should I update the usage after the release of Flex (in Alpha mode, I know but that's a major release) via the blog ? 

Flex change some configuration added via this PR and it's probably a good idea to show via the doc how to inject the ADR logic using Flex. 

Probably @javiereguiluz can help us to decide ?

HeahDude commented on controller/service.rst:
```diff
@@ -80,6 +80,98 @@ If your controller implements the ``__invoke()`` method - popular with the
 Action-Domain-Response (ADR) pattern, you can simply refer to the service id
 (``AppBundle\Controller\HelloController`` or ``app.hello_controller`` for example).
 
+If you want to use the ADR pattern rather than the default controller approach, you can use
```
I think this deserves its own article to be referenced here.

HeahDude commented on controller/service.rst:
```diff
@@ -80,6 +80,98 @@ If your controller implements the ``__invoke()`` method - popular with the
+            resource: '../../src/AppBundle/Action/'
+            public: true
+
+Once the file is updated, delete your Controller folder and create an Action one then 
```
"Once the file is updated, delete your Controller folder and create an Action class using the ADR principles, i.e ::"

HeahDude commented on controller/service.rst:
```diff
@@ -80,6 +80,98 @@ If your controller implements the ``__invoke()`` method - popular with the
+    namespace AppBundle\Action;
+
+    use Symfony\Component\HttpFoundation\Response;
+    use Symfony\Component\Templating\EngineInterface;
```
This component is to be deprecated, a Twig `Environment` should be used instead.

HeahDude commented on controller/service.rst:
```diff
@@ -80,6 +80,98 @@ If your controller implements the ``__invoke()`` method - popular with the
+
+    final class HelloAction
+    {
+        private $render;
```
Should be named `$twig` or `$renderer`.

HeahDude commented on controller/service.rst:
```diff
@@ -80,6 +80,98 @@ If your controller implements the ``__invoke()`` method - popular with the
+
+        // app/config/routing.php
+        $collection->add('hello', new Route('/hello', array(
+            '_controller' => 'AppBundle\Action\HelloAction',
```
`HelloAction::class`

HeahDude reviewed (commented):
Thank you for proposing this! Not sure this is the best place to do it though.

Guikingone commented:
@HeahDude Hi, thanks for the review, in fact, after reading your messages, I was thinking about a different approach, we all know that the doc should reflect the current usage of the framework, in this approach, I was thinking of proposing a new chapter in the "Guides" part, I don't have a specific name in mind but something like "Logic" or "Pattern" could be useful, in this chapter, I was thinking about linking to specific chapter about ADR, API or even MVC (using the Sf approach and actual code), we can build a specific API chapter like @javiereguiluz was talking about in the dedicated Slack then adding the ADR chapter, this way, we can allow the visitor to find and easily see how the pattern/logic can be adapted into Sf via example and dedicated tips, any thought about this ? 

For the corrections, I correct all your recommendations as soon as I'm back on my computer.

Guikingone committed:
Update service.rst

Updating the classes call and prepare for dedicated article.

Guikingone commented:
@HeahDude Just corrected the classes call and prepare for the dedicated article, gonna clone the repo on my computer and add the new file as soon as possible (probably before Sunday).

Guikingone committed:
Added the dedicated file and link from service.rst

Guikingone commented:
@HeahDude Hi, I've just added the dedicated file and the link from service.rst.

yceruto commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+    {
+        private $renderer;
+
+        public function __construct(Environment $render)
```
`$render` -> `$renderer`, same below.

yceruto replied:
Although it is usually called `$twig` for `Environment` instances.

HeahDude replied:
👍 for `$twig`.

yceruto commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+        }
+    }
+
+Final though
```
though**t**?

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+How to implement the ADR pattern
+================================
+
+In Symfony, you're used to implement the MVC pattern and extending the default 'Controller'
```
`'Controller'` =>
```rst
:class:`Symfony\\Bundle\\FrameworkBundle\\Controller\\Controller`
```

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+In Symfony, you're used to implement the MVC pattern and extending the default 'Controller'
+class, since the 3.3 update, Symfony is capable of using natively the ADR approach.
+
+Updating your internal logic
```
"Updating your configuration"?

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+================================
+
+In Symfony, you're used to implement the MVC pattern and extending the default 'Controller'
+class, since the 3.3 update, Symfony is capable of using natively the ADR approach.
```
I would put a dot before "since".

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+Updating your internal logic
+----------------------------
+
+As you saw from earlier example, you must update the services.yml file in order to
```
"As you saw from earlier example" does not mean anything is this context what introducing by:
```rst
You just need to update `the services.yml` file in order to auto configure your actions::
```
(Notice there is no space before the `:: `.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+
+    # ...
+
+    services:
```
I suggest to add xml and PHP versions too.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+    namespace AppBundle\Action;
+
+    use Twig\Environment;
+    use Symfony\Component\Templating\EngineInterface;
```
This use statement is no longer needed.

Guikingone committed:
Update adr.rst

Updated the twig call.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+            $this->render = $render;
+        }
+
+        public function __invoke()
```
You should add a return type hint.

Guikingone commented:
@yceruto Should be better, sorry for the errors.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+.. tip::
+
+    As described in the DependencyInjection doc, you must use the __construct() injection
+    approach, this way, your class is easier to update and keep in sync with the framework internal
```
"the framework" => "any framework"

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+    approach, this way, your class is easier to update and keep in sync with the framework internal
+    services.
+
+By default, we define the class with the final keyword because this class shouldn't been extended,
```
"been" => "be"

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+
+        // app/config/routing.php
+        $collection->add('hello', new Route('/hello', array(
+            '_controller' => 'HelloAction::class',
```
Please remove the quotes and add a use statement.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+    is faster to return and easier to keep out of the framework logic.
+
+
+Once this is done, you can define the routes like before using multiples approach :
```
In English there is no space before a colon `:`.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+Accessing the request
+---------------------
+
+As you can imagine, as the logic evolve, your class isn't capable of accessing
```
"isn't" => "is"

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+        }
+    }
+
+Like you can easily imagine, the container is the best option to gain access to ths request,
```
"container" =>
```rst
:class:`Symfony\\Component\\HttpFoundation\\RequestStack`
```

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,187 @@
+
+        public function __invoke(Request $request)
+        {
+            $data = $this->requestStack->getCurrentRequest()->get('id');
```
`$data` is not used, you should at least pass it to the view.

HeahDude commented on controller/service.rst:
```diff
@@ -80,6 +80,11 @@ If your controller implements the ``__invoke()`` method - popular with the
 (``AppBundle\Controller\HelloController`` or ``app.hello_controller`` for example).
 
+As this approach is evolving faster and can be easily transposed into Symfony, we recommend
+you to read the following part :
```
No space before colon.

HeahDude commented:
Thanks @Guikingone for your work on this!

Guikingone committed:
Update adr.rst

Updated the syntax errors and unused use statements.

Guikingone committed:
Update service.rst

Removed the space before colon.

Guikingone committed:
Update adr.rst

Add the use statement for routing and corrected the request part.

Guikingone commented:
@HeahDude Should be better, for the $data problem, I've passed the variable using the short syntax [] for the render method, is that correct or should I use the array() approach?

HeahDude commented on controller/adr.rst:
```diff
@@ -4,14 +4,15 @@ single: Action Domain Responder approach
 
-As you saw from earlier example, you must update the services.yml file in order to
-use the latest features of the DependencyInjection component, this way, here's the updates ::
+As the framework evolve, you must update the services.yml file in order to
```
"evolves"

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+    By using the final approach and the private visibility (inside the container), our class
+    is faster to return and easier to keep out of the framework logic.
+
+
```
Extra blank line.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+    is faster to return and easier to keep out of the framework logic.
+
+
+Once this is done, you can define the routes like before using multiples approach:
```
"multiples approach" => "multiple approaches"

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+            $this->twig = $twig;
+        }
+
+        public function __invoke()
```
`: Response`

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+
+        public function __invoke()
+        {
+            return new Response($this->twig->render('default/index.html.twig'));
```
The `Response` class is missing a use statement.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+
+        // ...
+
+        public function __invoke(Request $request)
```
Please also add a use statement for the `Request` in this example.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+        }
+    }
+
+Like you can easily imagine, the container is the best option to gain access to ths request,
```
"container" is confusing here and should be
```rst
:class:`Symfony\\Component\Httpfoundation\RequestStack`
```

typo "ths" => "the".

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+    use Twig\Environment;
+    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HttpFoundation\RequestStack;
+    use Symfony\Component\Templating\EngineInterface;
```
Should be removed.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+
+        public function __invoke(Request $request)
+        {
+            return new Response($this->twig->render('default/index.html.twig'));
```
I would add the `$data` example at this point and remove the other example below, that's too verbose.

Also I would remove the needless `Request` parameter here (along with its use statement).

Finally to keep the example working here a use statement should be added for the Response too.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+    }
+
+Final thought
+------------
```
Missing a dash.

HeahDude reviewed (commented):
Thanks again @Guikingone! I think some wording could be improved but I'll let our friend @javiereguiluz give his opinion on this matter.

HeahDude commented on controller/adr.rst:
```diff
@@ -0,0 +1,189 @@
+        {
+            $data = $this->requestStack->getCurrentRequest()->get('id');
+
+            return new Response($this->twig->render('default/index.html.twig', ['data' => $data]));
```
We still use the long syntax in the core and in the docs to ease merging olders branches.

Guikingone committed:
Update adr.rst

Update the missing use and adding the return type, also correct the typos

Guikingone committed:
Corrected the index typo.

Guikingone committed:
Update adr.rst

Add the backslash escape

Guikingone commented:
@HeahDude Hi, should be better for the missing use and typo problems.

theofidry commented on controller/adr.rst:
```diff
@@ -154,34 +153,16 @@ using this approach, a simple update is recommended ::
-
-            return new Response($this->twig->render('default/index.html.twig', ['data' => $data]));
+            
+            return new Response($this->twig->render('default/index.html.twig', array('data' => $data));
```
i think PHP7 is assumed in the doc

Guikingone replied:
Do you think we should use the [] syntax?

yceruto commented on controller/adr.rst:
```diff
@@ -34,6 +34,7 @@ Once the file is updated, delete your Controller folder and create an Action cla
     namespace AppBundle\Action;
 
     use Twig\Environment;
+    use Symfony\Component\HTTPFoundation\Response;
```
`HTTPFoundation` => `HttpFoundation`?

yceruto commented on controller/adr.rst:
```diff
@@ -122,25 +122,24 @@ the request from simple method injection like this ::
 
     use Twig\Environment;
-    use Symfony\Component\HttpFoundation\Request;
+    use Symfony\Component\HTTPFoundation\Response;
```
Same here.

yceruto commented on controller/adr.rst:
```diff
@@ -122,25 +122,24 @@ the request from simple method injection like this ::
         // ...
 
-        public function __invoke(Request $request)
+        public function __invoke(Request $request) : Response
```
The extra space before `:` could be removed?

Guikingone committed:
Update adr.rst

Updated the HttpFoundation calls.

Guikingone commented:
@yceruto Just corrected the HttpFoundation calls and removed the extra space before the response type.

xabbuh commented:
I am labeling this PR as "on hold" as @dunglas wanted to provide feedback after his holidays.

lyrixx commented on controller/adr.rst:
```diff
@@ -0,0 +1,170 @@
+        }
+    }
+
+Like you can easily imagine, the :class:`Symfony\\Component\\Httpfoundation\\RequestStack` is the best option to gain access to the request, using this approach, a simple update is recommended and the access to request parameters is way easier::
```
Why is it better ?

yceruto replied:
Using the `Request` from simple method injection still works in ADR patern, no sure it is the best sample :/

Guikingone commented:
@xabbuh No problem :)

@lyrixx In fact, you can use the Request class with the method injection but the __constructor approach is way better since she respect the best practices (injection via the constructor for better reusability and evolutivity) and allow for better code quality and simplicity, I know, the documentation teaches the inverse approach (via getter injection or method injection) but I think that the injection via the constructor still a way better approach for code structure.

@yceruto Yep, still work but like said before, the __construct injection stay cleaner and in accord with best practices, plus the documentation already told that the Request can be found using the RequestStack if you're in a service, as long as ADR is an evolution of MVC (and the fact that using the Controller class is a bad idea as soon as it make testing way harder), I think it's better to retrieve the Request from the RequestStack class and not from the Request class and get away from bad practices: [documentation](https://symfony.com/doc/current/service_container/request.html).

Not sure If I'm pretty clear but by using the Request injection (via __invoke arguments), you make testing harder (especially if you test the existence of an attribute within your class) and evolutivity harder again because of the method dependence with this special class, by using the __construct approach, the class stay dependent of the "Request" Layer but by using the RequestStack and not a method attribute.

yceruto commented:
> I think that the injection via the constructor still a way better approach for code structure.
> still work but like said before, the __construct injection stay cleaner and in accord with best practices

I'm no sure it should be applied to `RequestStack` and controllers where `Request` is not a service, but an argument resolved.

> allow for better code quality and simplicity

**Before:**
```php
private $requestStack;

public function __construct(RequestStack $requestStack)
{
    $this->requestStack = $requestStack;
}

public function __invoke()
{
    $request = $this->requestStack->getCurrentRequest();
    // ... do something with the request
}
```
**After:**
```php
public function __invoke(Request $request)
{
    // ... do something with the request
}
```

Not sure how is it true :P

> plus the documentation already told that the Request can be found using the RequestStack if you're in a service

Indeed, it's a bad practice ([LoD](https://en.wikipedia.org/wiki/Law_of_Demeter)), but the workaround for services (not controller services).

> I think it's better to retrieve the Request from the RequestStack class and not from the Request class and get away from bad practices
> by using the Request injection (via __invoke arguments), you make testing harder

I think the opposite, again ([LoD](https://en.wikipedia.org/wiki/Law_of_Demeter))

Guikingone commented:
@yceruto Oh, I see what you talking about, doesn't think about this vision, in fact, I was thinking about pure injection and not controller as service, in the fact of pure injection, constructor approach stays simpler to update and in line with the best practices.

lyrixx commented:
I think we don't have the same definition for simplicity nor "best-practice".

Guikingone commented:
@lyrixx In fact, maybe I've used the wrong terms, when I say that the constructor injection stay better and easier, it's in the way that for a new comer, the __construct injection stay easier to understand, even if you don't understand how the container works, if you understand the DI pattern, you can easily understand that the container find the right class and inject her into your class, that's a common approach (even in the online tutorials, the __constructor approach is largely used), that's why I say that it's easier to understand.

Plus, the method injection is "magic" in the way that you don't declare "publicly" that your class need something, you say that this method need the Request object and after refresh, Sf injects the object and voila, that like the auto wiring approach, even if everyone is using it now, I continue to say that it's a bad approach in order to help new comer to start using the framework. 

I'm gonna correct the PR in order to use the method injection :)

Guikingone committed:
Update adr.rst

Removed the RequestStack injection.

Guikingone committed:
Update adr.rst

Updated the Request call and the Twig\Environment render() call.

Guikingone commented:
Updated the Request call then the Twig render() call.

yceruto commented on controller/adr.rst:
```diff
@@ -0,0 +1,142 @@
+        public function __construct(Environment $twig)
+        {
+            $this->twig = $twig;
+        }
```
Thinking again about this approach, if we have just one method/controller, shouldn't be simpler inject the required dependencies into `__invoke()` directly? Thus, we are avoiding define one property per service argument, if so the action service need to be tagged with `controller.service_arguments`.

**After**
```php
final class HelloAction
{
    public function __invoke(Environment $twig): Response
    {
         return new Response($twig->render('default/index.html.twig'));
    }
}
```

Guikingone commented:
@yceruto It can work but this make appear the "controller" word inside a "pattern" who don't use it, for me, it can be strange for newcomer (and even for experimented developer) to use the word controller when your class became an Action.

This says, @dunglas can probably say more than me about this pattern.

Guikingone committed:
Update adr.rst

Corrected the service injection and services configuration.

Guikingone commented:
@yceruto Just made the corrections on service injection via the tags, sorry for the delay.

yceruto commented on controller/adr.rst:
```diff
@@ -0,0 +1,135 @@
+        AppBundle\Action\:
+            resource: '../../src/AppBundle/Action/'
+            tags: ['controller.service_arguments']
+            public: true
```
This behavior has been modified for 3.3, now all the services tagged with `controller.service_arguments` are public by definition and according to https://github.com/symfony/symfony-docs/pull/8415 we can remove it so.

yceruto commented on controller/adr.rst:
```diff
@@ -0,0 +1,135 @@
+As the framework evolve, you must update the services.yml file in order to
+use the latest features of the DependencyInjection component, this way, here's the updates::
+
+    # ...
```
can be removed?

Guikingone committed:
Update adr.rst

Guikingone committed:
Update adr.rst

Guikingone commented:
@yceruto Should be better.

Guikingone commented:
@HeahDude Your requested changes stay blocking while they were accepted and corrected, bug?

Guikingone committed:
[ADD](Responder classe)[!P2]

Added a Responder call and the configuration linked.

Guikingone commented:
Just added a new block for Responder injection and configuration.

Guikingone committed:
[FIX](Syntax)[!P2]

Fix the yaml call and the extra line in the responder configuration block.

dunglas commented on controller/adr.rst:
```diff
@@ -0,0 +1,204 @@
+
+In Symfony, you're used to implement the MVC pattern and extending the default :class:`Symfony\\Bundle\\FrameworkBundle\\Controller\\Controller`
+class.
+Since the 3.3 update, Symfony is capable of using natively the ADR approach.
```
Maybe it is worth it to add a link to https://github.com/pmjones/adr

Guikingone committed:
[FIX](ADR Repository)[!P2]

Just added the link to the official ADR repository.

Guikingone commented:
@dunglas Should be better.

fabpot commented on controller/adr.rst:
```diff
@@ -0,0 +1,206 @@
+How to implement the ADR pattern
+================================
+
+In Symfony, you're used to implement the MVC pattern and extending the default :class:`Symfony\\Bundle\\FrameworkBundle\\Controller\\Controller`
```
Symfony has never promoted the MVC pattern (at least, not since version 2). Instead, we are presenting Symfony as being a Request/Response framework where the controller converts a Request to a Response via a Controller.

Guikingone replied:
Yes, my apologies for this words, the meaning of my "text" was to clearly say that a lot of developers use Symfony with MVC and that using this pattern is way more used than just transforming a Request into Response (which is handled by the framework in a certain way, the actual transformation occurs in the controller).

fabpot commented on controller/adr.rst:
```diff
@@ -0,0 +1,206 @@
+            tags: 
+                - 'controller.service_arguments'
+
+Now that the container knows about our actions, time to build a simple Action !
```
extra whitespace before the `!`

fabpot commented on controller/adr.rst:
```diff
@@ -0,0 +1,206 @@
+Updating your classes
+---------------------
+
+As the framework evolve, you must update your classes, first, delete your Controller folder and create an Actions one then a new class using the ADR principles, for this example, call it ``HelloAction.php``: 
```
typo: `evolves`

fabpot commented on controller/adr.rst:
```diff
@@ -0,0 +1,206 @@
+Keep in mind that this approach can be completely different from what you're used to use, in order to
+keep your code clean and easy to maintain, we recommend to use this approach only if your code is
+decoupled from the internal framework logic (like with Clean Architecture approach) or if you start a new
+project and need to keep the logic linked to your business rules.
```
I don't understand this last paragraph and the supposed benefits. I don't see how this approach or what you refer to the old one is cleaner or easier. I don't understand why this approach is more decoupled than the other one. I'm not saying this is good or bad, just that from a beginner's perspective, the benefits are not very clear.

fabpot commented on controller/service.rst:
```diff
@@ -80,6 +80,11 @@ If your controller implements the ``__invoke()`` method - popular with the
 Action-Domain-Response (ADR) pattern, you can simply refer to the service id
 (``AppBundle\Controller\HelloController`` or ``app.hello_controller`` for example).
 
+As this approach is evolving faster and can be easily transposed into Symfony, we recommend
```
How is it evolving?

Guikingone replied:
At this stage, not so much, like before, there's was an error in the real sense of my words, ADR still young and the "logic" keep evolving with time and usage by the developers, the true meaning was to say that using ADR still a "WIP" in Symfony and as the "logic" of ADR still evolving by the feedbacks of the developers, his implementation inside Symfony can be modified with future evolution of the framework or pattern. 

_In fact, it was possible since Symfony 3.1 when recompiling the container (or like @dunglas do it, by injecting the classes as controllers and mapping the return of each one of them) but using it in production is way easier since 3.3 and 3.4 (and even easier in 4.0)_

Guikingone committed:
fix(typo): fix on typo and space.

javiereguiluz commented:
What should we do with this proposal? I don't see people talking or promoting ADR on Symfony Slack or blogs or web sites or aggregators like Hacker News. Should we close without merging?

Guikingone commented:
For me, as ADR is only an evolution of MVC (which is mainly used in today's applications), I think that we can close this PR, it's not the role of Symfony to promote this type of architecture as everyone can have a different point of view when it comes to how to implement it 🙂

fabpot commented:
I agree that this is opinionated and should not be part of the official docs.

javiereguiluz commented:
OK, let's close then. Guillaume thanks for being so understanding. Thanks.
